Flaky test: TestClientStreamingError

Data race:
<details><summary>Log data</summary>

```

E0628 01:04:11.039628    6381 server.go:841] grpc: server failed to encode response:  rpc error: code = Internal desc = grpc: error while marshaling: 3987^12 + 4365^12 = 4472^12
E0628 01:04:11.055868    6381 server.go:841] grpc: server failed to encode response:  rpc error: code = Internal desc = grpc: error while marshaling: 3987^12 + 4365^12 = 4472^12
E0628 01:04:11.061755    6381 server.go:841] grpc: server failed to encode response:  rpc error: code = Internal desc = grpc: error while marshaling: 3987^12 + 4365^12 = 4472^12
E0628 01:04:11.077099    6381 server.go:841] grpc: server failed to encode response:  rpc error: code = Internal desc = grpc: error while marshaling: 3987^12 + 4365^12 = 4472^12
E0628 01:04:11.093710    6381 server.go:841] grpc: server failed to encode response:  rpc error: code = Internal desc = grpc: error while marshaling: 3987^12 + 4365^12 = 4472^12
E0628 01:04:11.111580    6381 server.go:841] grpc: server failed to encode response:  rpc error: code = Internal desc = grpc: error while marshaling: 3987^12 + 4365^12 = 4472^12
E0628 01:04:14.424052    6381 http2_server.go:771] transport: failed to marshal rpc status: code:13 message:"\377\376\375" details:<type_url:"type.googleapis.com/grpc.testing.Empty" > , error: proto: string field "google.rpc.Status.Message" contains invalid UTF-8
==================
WARNING: DATA RACE
Write at 0x00c42001a4e0 by goroutine 52:
  google.golang.org/grpc/transport.(*http2Client).closeStream()
      /home/travis/gopath/src/google.golang.org/grpc/transport/http2_client.go:646 +0x82
  google.golang.org/grpc/transport.(*http2Client).operateHeaders()
      /home/travis/gopath/src/google.golang.org/grpc/transport/http2_client.go:1106 +0x51e
  google.golang.org/grpc/transport.(*http2Client).reader()
      /home/travis/gopath/src/google.golang.org/grpc/transport/http2_client.go:1163 +0x98e

Previous read at 0x00c42001a4e0 by goroutine 21:
  google.golang.org/grpc.(*clientStream).shouldRetry()
      /home/travis/gopath/src/google.golang.org/grpc/transport/transport.go:339 +0x29f
  google.golang.org/grpc.(*clientStream).retryLocked()
      /home/travis/gopath/src/google.golang.org/grpc/stream.go:505 +0x8f
  google.golang.org/grpc.(*clientStream).withRetry()
      /home/travis/gopath/src/google.golang.org/grpc/stream.go:546 +0x165
  google.golang.org/grpc.(*clientStream).CloseSend()
      /home/travis/gopath/src/google.golang.org/grpc/stream.go:665 +0xfa
  google.golang.org/grpc/test/grpc_testing.(*testServiceStreamingInputCallClient).CloseAndRecv()
      /home/travis/gopath/src/google.golang.org/grpc/test/grpc_testing/test.pb.go:633 +0x52
  google.golang.org/grpc/test.testClientStreamingError()
      /home/travis/gopath/src/google.golang.org/grpc/test/end2end_test.go:3757 +0x688
  google.golang.org/grpc/test.TestClientStreamingError()
      /home/travis/gopath/src/google.golang.org/grpc/test/end2end_test.go:3727 +0x189
  testing.tRunner()
      /home/travis/.gimme/versions/go1.10.3.linux.amd64/src/testing/testing.go:777 +0x16d

Goroutine 52 (running) created at:
  google.golang.org/grpc/transport.newHTTP2Client()
      /home/travis/gopath/src/google.golang.org/grpc/transport/http2_client.go:265 +0x1114
  google.golang.org/grpc/transport.NewClientTransport()
      /home/travis/gopath/src/google.golang.org/grpc/transport/transport.go:506 +0xca
  google.golang.org/grpc.(*addrConn).createTransport()
      /home/travis/gopath/src/google.golang.org/grpc/clientconn.go:1334 +0x3fc
  google.golang.org/grpc.(*addrConn).resetTransport()
      /home/travis/gopath/src/google.golang.org/grpc/clientconn.go:1292 +0x782
  google.golang.org/grpc.(*addrConn).connect.func1()
      /home/travis/gopath/src/google.golang.org/grpc/clientconn.go:991 +0x3c

Goroutine 21 (running) created at:
  testing.(*T).Run()
      /home/travis/.gimme/versions/go1.10.3.linux.amd64/src/testing/testing.go:824 +0x564
  testing.runTests.func1()
      /home/travis/.gimme/versions/go1.10.3.linux.amd64/src/testing/testing.go:1063 +0xa4
  testing.tRunner()
      /home/travis/.gimme/versions/go1.10.3.linux.amd64/src/testing/testing.go:777 +0x16d
  testing.runTests()
      /home/travis/.gimme/versions/go1.10.3.linux.amd64/src/testing/testing.go:1061 +0x4e1
  testing.(*M).Run()
      /home/travis/.gimme/versions/go1.10.3.linux.amd64/src/testing/testing.go:978 +0x2cd
  main.main()
      _testmain.go:280 +0x22a
==================
==================
WARNING: DATA RACE
Write at 0x00c42001a4c8 by goroutine 52:
  google.golang.org/grpc/transport.(*http2Client).closeStream()
      /home/travis/gopath/src/google.golang.org/grpc/transport/http2_client.go:648 +0x4c9
  google.golang.org/grpc/transport.(*http2Client).operateHeaders()
      /home/travis/gopath/src/google.golang.org/grpc/transport/http2_client.go:1106 +0x51e
  google.golang.org/grpc/transport.(*http2Client).reader()
      /home/travis/gopath/src/google.golang.org/grpc/transport/http2_client.go:1163 +0x98e

Previous read at 0x00c42001a4c8 by goroutine 21:
  google.golang.org/grpc/transport.(*Stream).Trailer()
      /home/travis/gopath/src/google.golang.org/grpc/transport/transport.go:306 +0x3f
  google.golang.org/grpc.(*clientStream).shouldRetry()
      /home/travis/gopath/src/google.golang.org/grpc/stream.go:438 +0xa55
  google.golang.org/grpc.(*clientStream).retryLocked()
      /home/travis/gopath/src/google.golang.org/grpc/stream.go:505 +0x8f
  google.golang.org/grpc.(*clientStream).withRetry()
      /home/travis/gopath/src/google.golang.org/grpc/stream.go:546 +0x165
  google.golang.org/grpc.(*clientStream).CloseSend()
      /home/travis/gopath/src/google.golang.org/grpc/stream.go:665 +0xfa
  google.golang.org/grpc/test/grpc_testing.(*testServiceStreamingInputCallClient).CloseAndRecv()
      /home/travis/gopath/src/google.golang.org/grpc/test/grpc_testing/test.pb.go:633 +0x52
  google.golang.org/grpc/test.testClientStreamingError()
      /home/travis/gopath/src/google.golang.org/grpc/test/end2end_test.go:3757 +0x688
  google.golang.org/grpc/test.TestClientStreamingError()
      /home/travis/gopath/src/google.golang.org/grpc/test/end2end_test.go:3727 +0x189
  testing.tRunner()
      /home/travis/.gimme/versions/go1.10.3.linux.amd64/src/testing/testing.go:777 +0x16d

Goroutine 52 (running) created at:
  google.golang.org/grpc/transport.newHTTP2Client()
      /home/travis/gopath/src/google.golang.org/grpc/transport/http2_client.go:265 +0x1114
  google.golang.org/grpc/transport.NewClientTransport()
      /home/travis/gopath/src/google.golang.org/grpc/transport/transport.go:506 +0xca
  google.golang.org/grpc.(*addrConn).createTransport()
      /home/travis/gopath/src/google.golang.org/grpc/clientconn.go:1334 +0x3fc
  google.golang.org/grpc.(*addrConn).resetTransport()
      /home/travis/gopath/src/google.golang.org/grpc/clientconn.go:1292 +0x782
  google.golang.org/grpc.(*addrConn).connect.func1()
      /home/travis/gopath/src/google.golang.org/grpc/clientconn.go:991 +0x3c

Goroutine 21 (running) created at:
  testing.(*T).Run()
      /home/travis/.gimme/versions/go1.10.3.linux.amd64/src/testing/testing.go:824 +0x564
  testing.runTests.func1()
      /home/travis/.gimme/versions/go1.10.3.linux.amd64/src/testing/testing.go:1063 +0xa4
  testing.tRunner()
      /home/travis/.gimme/versions/go1.10.3.linux.amd64/src/testing/testing.go:777 +0x16d
  testing.runTests()
      /home/travis/.gimme/versions/go1.10.3.linux.amd64/src/testing/testing.go:1061 +0x4e1
  testing.(*M).Run()
      /home/travis/.gimme/versions/go1.10.3.linux.amd64/src/testing/testing.go:978 +0x2cd
  main.main()
      _testmain.go:280 +0x22a
==================
--- FAIL: TestClientStreamingError (0.08s)
	end2end_test.go:538: Running test in tcp-clear-v1-balancer environment...
	end2end_test.go:538: Running test in tcp-tls-v1-balancer environment...
	end2end_test.go:538: Running test in tcp-clear environment...
	end2end_test.go:538: Running test in tcp-tls environment...
	end2end_test.go:538: Running test in no-balancer environment...
	testing.go:730: race detected during execution of test
E0628 01:05:31.448787    6381 server.go:841] grpc: server failed to encode response:  rpc error: code = Internal desc = grpc: error while marshaling: 3987^12 + 4365^12 = 4472^12
E0628 01:05:31.477227    6381 server.go:841] grpc: server failed to encode response:  rpc error: code = Internal desc = grpc: error while marshaling: 3987^12 + 4365^12 = 4472^12
E0628 01:05:31.484234    6381 server.go:841] grpc: server failed to encode response:  rpc error: code = Internal desc = grpc: error while marshaling: 3987^12 + 4365^12 = 4472^12
E0628 01:05:31.504712    6381 server.go:841] grpc: server failed to encode response:  rpc error: code = Internal desc = grpc: error while marshaling: 3987^12 + 4365^12 = 4472^12
E0628 01:05:31.527523    6381 server.go:841] grpc: server failed to encode response:  rpc error: code = Internal desc = grpc: error while marshaling: 3987^12 + 4365^12 = 4472^12
E0628 01:05:31.554799    6381 server.go:841] grpc: server failed to encode response:  rpc error: code = Internal desc = grpc: error while marshaling: 3987^12 + 4365^12 = 4472^12
E0628 01:05:35.143316    6381 http2_server.go:771] transport: failed to marshal rpc status: code:13 message:"\377\376\375" details:<type_url:"type.googleapis.com/grpc.testing.Empty" > , error: proto: string field "google.rpc.Status.Message" contains invalid UTF-8
```
</details>

--- a/stream.go
+++ b/stream.go
@@ -413,6 +413,9 @@ func (cs *clientStream) shouldRetry(err error) error {
 		// RPC is finished or committed; cannot retry.
 		return err
 	}
+	if cs.attempt.s != nil {
+		<-cs.attempt.s.Done()
+	}
 	if cs.firstAttempt && !cs.callInfo.failFast && (cs.attempt.s == nil || cs.attempt.s.Unprocessed()) {
 		// First attempt, wait-for-ready, stream unprocessed: transparently retry.
 		cs.firstAttempt = false
@@ -502,11 +505,11 @@ func (cs *clientStream) shouldRetry(err error) error {
 // Returns nil if a retry was performed and succeeded; error otherwise.
 func (cs *clientStream) retryLocked(lastErr error) error {
 	for {
+		cs.attempt.finish(lastErr)
 		if err := cs.shouldRetry(lastErr); err != nil {
 			cs.commitAttemptLocked()
 			return err
 		}
-		cs.attempt.finish(lastErr)
 		if err := cs.newAttemptLocked(nil, traceInfo{}); err != nil {
 			return err
 		}
@@ -796,6 +799,7 @@ func (a *csAttempt) recvMsg(m interface{}) (err error) {
 func (a *csAttempt) finish(err error) {
 	a.mu.Lock()
 	if a.finished {
+		a.mu.Unlock()
 		return
 	}
 	a.finished = true

--- a/stream.go
+++ b/stream.go
@@ -645,8 +645,8 @@ func (cs *clientStream) SendMsg(m interface{}) (err error) {
 func (cs *clientStream) RecvMsg(m interface{}) error {
 	err := cs.withRetry(func(a *csAttempt) error {
 		err := a.recvMsg(m)
-		if err != nil || !cs.desc.ServerStreams {
-			// err != nil or non-server-streaming indicates end of stream.
+		if !cs.desc.ServerStreams {
+			// Finish for non-server-streaming RPCs even on non-nil error.
 			a.finish(err)
 		}
 		return err

--- a/stream.go
+++ b/stream.go
@@ -413,6 +413,7 @@ func (cs *clientStream) shouldRetry(err error) error {
 		// RPC is finished or committed; cannot retry.
 		return err
 	}
+	// Wait for the trailers.
 	if cs.attempt.s != nil {
 		<-cs.attempt.s.Done()
 	}

--- a/stream.go
+++ b/stream.go
@@ -644,12 +644,7 @@ func (cs *clientStream) SendMsg(m interface{}) (err error) {
 
 func (cs *clientStream) RecvMsg(m interface{}) error {
 	err := cs.withRetry(func(a *csAttempt) error {
-		err := a.recvMsg(m)
-		if !cs.desc.ServerStreams {
-			// Finish for non-server-streaming RPCs even on non-nil error.
-			a.finish(err)
-		}
-		return err
+		return a.recvMsg(m)
 	}, cs.commitAttemptLocked)
 	if err != nil || !cs.desc.ServerStreams {
 		// err != nil or non-server-streaming indicates end of stream.

